File: go/technical/technical.go
package technical

import "math"

type MovingAverages struct {
	MA20  *float64
	MA50  *float64
	MA200 *float64
}

type DashboardMetrics struct {
	RSI      float64
	MA20     *float64
	MA50     *float64
	MA200    *float64
	Momentum float64
	Signal   string
	Score    float64
}

func MA(prices []float64, period int) float64 {
	n := len(prices)
	if n < period {
		return 0
	}
	sum := 0.0
	for i := n - period; i < n; i++ {
		sum += prices[i]
	}
	return sum / float64(period)
}

func CalculateMovingAverages(closes []float64) MovingAverages {
	var ma MovingAverages
	if len(closes) >= 20 {
		v := MA(closes, 20)
		ma.MA20 = &v
	}
	if len(closes) >= 50 {
		v := MA(closes, 50)
		ma.MA50 = &v
	}
	if len(closes) >= 200 {
		v := MA(closes, 200)
		ma.MA200 = &v
	}
	return ma
}

func RSI(closes []float64, period int) float64 {
	if len(closes) < period+1 {
		return 50
	}
	var gains, losses float64
	for i := 1; i <= period; i++ {
		delta := closes[i] - closes[i-1]
		if delta > 0 {
			gains += delta
		} else {
			losses -= delta
		}
	}
	avgGain := gains / float64(period)
	avgLoss := losses / float64(period)

	for i := period + 1; i < len(closes); i++ {
		delta := closes[i] - closes[i-1]
		if delta > 0 {
			avgGain = (avgGain*float64(period-1) + delta) / float64(period)
			avgLoss = (avgLoss * float64(period-1)) / float64(period)
		} else {
			avgGain = (avgGain * float64(period-1)) / float64(period)
			avgLoss = (avgLoss*float64(period-1) - delta) / float64(period)
		}
	}

	if avgLoss == 0 {
		return 100
	}
	rs := avgGain / avgLoss
	return 100 - (100 / (1 + rs))
}

func Momentum(closes []float64, period int) float64 {
	if len(closes) <= period {
		return 0
	}
	current := closes[len(closes)-1]
	past := closes[len(closes)-1-period]
	if past == 0 {
		return 0
	}
	return ((current - past) / past) * 100
}

func Signal(price float64, ma20, ma50 *float64, rsi float64) string {
	if ma20 == nil || ma50 == nil {
		return "HOLD"
	}
	if price > *ma20 && *ma20 > *ma50 && rsi < 70 {
		return "BUY"
	}
	if price < *ma20 && *ma20 < *ma50 && rsi > 30 {
		return "SELL"
	}
	return "HOLD"
}

func Score(buySignal, sellSignal bool, rsi float64) float64 {
	score := 50.0
	if buySignal {
		score += 25
	}
	if sellSignal {
		score -= 25
	}
	if rsi < 30 {
		score += 10
	}
	if rsi > 70 {
		score -= 10
	}
	return math.Max(0, math.Min(100, score))
}

func Dashboard(closes []float64) DashboardMetrics {
	ma := CalculateMovingAverages(closes)
	rsi := RSI(closes, 14)
	mom := Momentum(closes, 20)

	currentPrice := 0.0
	if len(closes) > 0 {
		currentPrice = closes[len(closes)-1]
	}

	sig := Signal(currentPrice, ma.MA20, ma.MA50, rsi)
	buy := sig == "BUY"
	sell := sig == "SELL"

	return DashboardMetrics{
		RSI: rsi, MA20: ma.MA20, MA50: ma.MA50, MA200: ma.MA200,
		Momentum: mom, Signal: sig, Score: Score(buy, sell, rsi),
	}
}

File: python/lotusmarket/technical.py
"""Technical analysis indicators for Vietnamese stocks."""

import numpy as np
import pandas as pd
from dataclasses import dataclass
from typing import Optional


@dataclass
class DashboardMetrics:
    rsi: float = 0.0
    ma20: Optional[float] = None
    ma50: Optional[float] = None
    ma200: Optional[float] = None
    momentum: float = 0.0
    signal: str = "HOLD"
    score: float = 50.0


def ma(series: pd.Series, period: int) -> pd.Series:
    """Simple moving average."""
    return series.rolling(window=period, min_periods=period).mean()


def ma_value(prices: pd.Series, period: int) -> Optional[float]:
    """Latest MA value, or None if insufficient data."""
    if len(prices) < period:
        return None
    return float(prices.iloc[-period:].mean())


def rsi(series: pd.Series, period: int = 14) -> pd.Series:
    """RSI using Wilder's smoothing. Returns Series."""
    delta = series.diff()
    gain = delta.where(delta > 0, 0.0)
    loss = (-delta).where(delta < 0, 0.0)
    avg_gain = gain.ewm(alpha=1 / period, min_periods=period, adjust=False).mean()
    avg_loss = loss.ewm(alpha=1 / period, min_periods=period, adjust=False).mean()
    rs = avg_gain / avg_loss
    return 100 - (100 / (1 + rs))


def rsi_value(prices: pd.Series, period: int = 14) -> float:
    """Latest RSI value. Returns 50 if insufficient data."""
    if len(prices) < period + 1:
        return 50.0
    r = rsi(prices, period)
    val = r.iloc[-1]
    if pd.isna(val):
        return 50.0
    return float(val)


def momentum(prices: pd.Series, period: int = 20) -> float:
    """Rate of change over N periods as percentage."""
    if len(prices) <= period:
        return 0.0
    current = float(prices.iloc[-1])
    past = float(prices.iloc[-1 - period])
    if past == 0:
        return 0.0
    return ((current - past) / past) * 100


def signal(prices: pd.Series) -> str:
    """BUY/SELL/HOLD based on MA crossover + RSI zones."""
    ma20 = ma_value(prices, 20)
    ma50 = ma_value(prices, 50)
    if ma20 is None or ma50 is None:
        return "HOLD"
    price = float(prices.iloc[-1])
    r = rsi_value(prices, 14)
    if price > ma20 > ma50 and r < 70:
        return "BUY"
    if price < ma20 < ma50 and r > 30:
        return "SELL"
    return "HOLD"


def score(prices: pd.Series) -> float:
    """Composite 0-100 score."""
    sig = signal(prices)
    r = rsi_value(prices, 14)
    s = 50.0
    if sig == "BUY":
        s += 25
    elif sig == "SELL":
        s -= 25
    if r < 30:
        s += 10
    if r > 70:
        s -= 10
    return max(0, min(100, s))


def dashboard(prices: pd.Series) -> DashboardMetrics:
    """All-in-one technical assessment."""
    return DashboardMetrics(
        rsi=rsi_value(prices, 14),
        ma20=ma_value(prices, 20),
        ma50=ma_value(prices, 50),
        ma200=ma_value(prices, 200),
        momentum=momentum(prices, 20),
        signal=signal(prices),
        score=score(prices),
    )


def aggregate_weekly(df: pd.DataFrame) -> pd.DataFrame:
    """Convert daily OHLCV DataFrame to weekly candles.
    Input must have columns: date, open, high, low, close, volume.
    """
    if df.empty:
        return pd.DataFrame()
    d = df.copy()
    d["date"] = pd.to_datetime(d["date"])
    d = d.set_index("date").sort_index()
    weekly = (
        d.resample("W")
        .agg(
            {
                "open": "first",
                "high": "max",
                "low": "min",
                "close": "last",
                "volume": "sum",
            }
        )
        .dropna()
    )
    weekly = weekly.reset_index()
    weekly["date"] = weekly["date"].dt.strftime("%Y-%m-%d")
    return weekly
